### DISCUSSION TITLE
render HTML and inject data for SPA

### DISCUSSION THREAD

**Discussion by lMINERl** [vikejs/vike#1490] [Discussion] [answered]
I'm trying to render html with  CSP 'nonce+{RNG}' in html Tag and on the scripts, stylesheets on SPA SolidJS app so searching for away to renderHtml with tags and option i need on each page reload.
Its my day 2 , while it said to be a bug I really doubt I'm on the right direction. 
![image](https://github.com/vikejs/vike/assets/19747719/262e47a6-740e-4bc9-b278-9a78e741ba43)

so when i add onRenderHtml to pages folder config `src/pages/+config.h.ts` it breaks
![image](https://github.com/vikejs/vike/assets/19747719/d47f6378-a3d1-4658-b5a1-832f2f011b5b)
`src/renderer/+onRenderHtml.ts`
![image](https://github.com/vikejs/vike/assets/19747719/5cf4aa55-8b3d-4048-a457-5cc423decb47)
`src/pages/index/+Page.tsx`
![image](https://github.com/vikejs/vike/assets/19747719/3682cd2c-618d-4cdd-8119-caa8fb0a1152)

I appreciate if someone could point me to right direction or what I'm doing wrong here , thanks in advance

**Comment by brillout** [comment]
Can you publish a reproduction for the error? (Although I don't think the error is related to what you're trying to achieve.)

**Reply by lMINERl** [reply]
1- setup a solid project `npx degit solidjs/templates/ts my-app`
2- add dependencies `pnpm add -D vike vikeSolid`
3- in `vite.config.js` comment `solidPlugin` and add `vike()` and `VikeSolid` from `vike/plugin` and `vike-solid/vite`
4- add `src/pages/index/+Page.tsx` file that has solidjs' entry app 
``` javascript
import { ClientOnly } from "vike-solid/ClientOnly";
function Index() {
  console.log("server side");
  
  return (
    <h1>
    server side
      <ClientOnly
        fallback={<h1>...loading</h1>}
        load={() => import("../App")}
        children={(Item) => {
          return <Item />;
        }}
      ></ClientOnly>
    </h1>
  );
}

export default Index;
```
5-  add `src/renderer/+onRenderHtml.tsx` 
``` javascript
// https://vike.dev/onRenderHtml

import { renderToStream, renderToString } from "solid-js/web";
import { escapeInject } from "vike/server";
import App from "../pages/App";
import type { OnRenderHtmlAsync } from "vike/types";
import { createUniqueId } from "solid-js";

const onRenderHtml: OnRenderHtmlAsync = async (
  pageContext,
): ReturnType<OnRenderHtmlAsync> => {
  const { Page } = pageContext;

  const stream = renderToStream(() => Page, {
    nonce: "123",
    renderId: createUniqueId(),
  });
  console.log(stream);

  const doc = escapeInject`<!DOCTYPE html>
    <html>
    <head>
       <style> .s{ background-color : red; }</style>
    </head>
      <body>
        <div id="page-view">${stream}</div>
      </body>
    </html>`;

  return doc;
};

export { onRenderHtml };
```
6- add `pages/+config.h.ts` with and add `onRenderHtml` to the config
``` javascript
import type { Config } from "vike/types";
import vikeSolidConfig from "vike-solid/config";
import { onRenderHtml } from "../renderer/+onRenderHtml";
import { renderToStream, renderToString } from "solid-js/web";
import { escapeInject } from "vike/server";

export default {
  extends: vikeSolidConfig,
  clientRouting: true,
  onRenderHtml: onRenderHtml,
} satisfies Config;
```
`pnpm start` it will build successfully go to browser `localhost:3000/` and refresh the following error  is displayed and you are presented browser with white screen `cannot GET/`

```
12:00:58 PM [vike][request(1)] HTTP request: /
12:00:58 PM [vike@0.4.161][request(1)][Bug] You stumbled upon a Vike bug. Go to https://github.com/vikejs/vike/issues/new and copy-paste this error. A maintainer will fix the bug (usually under 24 hours).
    at serializeConfigValueImported (file:///C:/Users/seect/Documents/general-reviewes/node_modules/.pnpm/vike@0.4.161_vite@4.5.2/node_modules/vike/dist/esm/shared/page-configs/serialize/serializeConfigValue.js:33:9)
    at file:///C:/Users/seect/Documents/general-reviewes/node_modules/.pnpm/vike@0.4.161_vite@4.5.2/node_modules/vike/dist/esm/node/plugin/plugins/importUserCode/v1-design/getVirtualFilePageConfigValuesAll.js:63:23
    at Array.forEach (<anonymous>)
    at getConfigValuesImported (file:///C:/Users/seect/Documents/general-reviewes/node_modules/.pnpm/vike@0.4.161_vite@4.5.2/node_modules/vike/dist/esm/node/plugin/plugins/importUserCode/v1-design/getVirtualFilePageConfigValuesAll.js:55:51)
    at getLoadConfigValuesAll (file:///C:/Users/seect/Documents/general-reviewes/node_modules/.pnpm/vike@0.4.161_vite@4.5.2/node_modules/vike/dist/esm/node/plugin/plugins/importUserCode/v1-design/getVirtualFilePageConfigValuesAll.js:37:16)
    at getVirtualFilePageConfigValuesAll (file:///C:/Users/seect/Documents/general-reviewes/node_modules/.pnpm/vike@0.4.161_vite@4.5.2/node_modules/vike/dist/esm/node/plugin/plugins/importUserCode/v1-design/getVirtualFilePageConfigValuesAll.js:28:18)
    at Context.load (file:///C:/Users/seect/Documents/general-reviewes/node_modules/.pnpm/vike@0.4.161_vite@4.5.2/node_modules/vike/dist/esm/node/plugin/plugins/importUserCode/index.js:50:30)
    at Object.load (file:///C:/Users/seect/Documents/general-reviewes/node_modules/.pnpm/vite@4.5.2/node_modules/vite/dist/node/chunks/dep-52909643.js:44323:32)
    at loadAndTransform (file:///C:/Users/seect/Documents/general-reviewes/node_modules/.pnpm/vite@4.5.2/node_modules/vite/dist/node/chunks/dep-52909643.js:54954:24)
    at instantiateModule (file:///C:/Users/seect/Documents/general-reviewes/node_modules/.pnpm/vite@4.5.2/node_modules/vite/dist/node/chunks/dep-52909643.js:55951:10)
12:00:58 PM [vike][request(1)][Warning] No error page found, we recommend defining an error page, see https://vike.dev/error-page
12:00:58 PM [vike][request(1)] HTTP response / ERR
```

**Reply by brillout** [reply]
Neat, please publish it as a GitHub repository I'll have a look at it then.

As for your original question, let me know if you're still having issues.

**Reply by lMINERl** [reply]
oh sure , sorry I didn't know you want the repo [https://github.com/lMINERl/general-reviewes](https://github.com/lMINERl/general-reviewes) and about my issue if you can point me what to do to have csp policy nonce-{RNG} on script tags and styles its from you and I'll be grateful

**Reply by brillout** [reply, marked as answer]
👍 I'll have a look at it later today.

> if you can point me what to do to have csp policy nonce-{RNG} on script tags and styles its from you and I'll be grateful

One thing you could do is to manipulate the generated HTML `pageContext.httpResponse.body` at your `renderPage()` server integration. Would that be an option for you?

**Reply by lMINERl** [reply]
anything works XD  thanks alot that will do

**Reply by brillout** [reply]
> I'll have a look at it later today.

I'll do it after I'm done with my current PR. I guess it isn't a blocker, but let me know if it is.

**Reply by brillout** [reply]
Fixed & pre-released as `0.4.162-commit-421de1c`.
